# Анализ крупного сервиса для чтения книг по подписке

### Введение

Нашему вниманию предоставлены данные крупного сервиса для чтения книг по подписке. Необходимо проанализировать базу данных.
В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


   **Описание данных**

Таблица **books**

Содержит данные о книгах:

- book_id — идентификатор книги;

- author_id — идентификатор автора;

- title — название книги;

- num_pages — количество страниц;

- publication_date — дата публикации книги;

- publisher_id — идентификатор издателя.

Таблица **authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers**
Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица **ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews**

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.


   **План**

1. Изучить общую информацию данных.


2. Подготовить по одному SQL-запросу для решения каждого задания.


3. Сформулировать выводы.

### Исследование данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

*Выведем первые 5 строк каждой таблицы, а затем основную информацию по ней.*

In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
books = pd.io.sql.read_sql(sql=text(query), con = con)
print('\t\t\t\t Данные о книгах')
display(books)

				 Данные о книгах


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
#выведем общую информацию по таблице:
query = '''SELECT * FROM books'''
con=engine.connect()
books = pd.io.sql.read_sql(sql=text(query), con = con)

print ('\n\tИнформация о books: \n') 
books.info()

#выведем размер исходной таблицы:
print ('\nКоличество строк и столбцов:', books.shape)

#проверим на пропуски
print ('\nИнформация по пропускам: \n') 
print(books.isnull().sum())

#проверим на дубликаты
print ('Количество дубликатов:', books.duplicated().sum())


	Информация о books: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB

Количество строк и столбцов: (1000, 6)

Информация по пропускам: 

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64
Количество дубликатов: 0


**Таблица books содержит данные о названии книги, количестве страниц, дате публикации, id книги, автора и издательства. В таблице 1000 строк и 6 столбцов. Дубликаты и пропущенные значения не обнаружены.**

In [4]:
query_2 = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
authors = pd.io.sql.read_sql(sql=text(query_2), con = con)
print('\t\t\t\t Данные об авторах')
display(authors)

				 Данные об авторах


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
#выведем общую информацию по таблице:
query_2 = '''SELECT * FROM authors'''
con=engine.connect()
authors = pd.io.sql.read_sql(sql=text(query_2), con = con)

print ('\n\tИнформация об authors: \n') 
authors.info()

#выведем размер исходной таблицы:
print ('\nКоличество строк и столбцов:', authors.shape)

#проверим на пропуски
print ('\nИнформация по пропускам: \n') 
print(authors.isnull().sum())

#проверим на дубликаты
print ('Количество дубликатов:', authors.duplicated().sum())


	Информация об authors: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB

Количество строк и столбцов: (636, 2)

Информация по пропускам: 

author_id    0
author       0
dtype: int64
Количество дубликатов: 0


**Таблица authors содержит данные об авторе и его id. В таблице 636 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены.**

In [6]:
query_3 = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
publishers = pd.io.sql.read_sql(sql=text(query_3), con = con)
print('\t\t\t\t Данные об издательствах')

				 Данные об издательствах


In [7]:
#выведем общую информацию по таблице:
query_3 = '''SELECT * FROM publishers'''
con=engine.connect()
publishers = pd.io.sql.read_sql(sql=text(query_3), con = con)

print ('\n\tИнформация о publishers: \n') 
publishers.info()

#выведем размер исходной таблицы:
print ('\nКоличество строк и столбцов:', publishers.shape)

#проверим на пропуски
print ('\nИнформация по пропускам: \n') 
print(publishers.isnull().sum())

#проверим на дубликаты
print ('Количество дубликатов:', publishers.duplicated().sum())


	Информация о publishers: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB

Количество строк и столбцов: (340, 2)

Информация по пропускам: 

publisher_id    0
publisher       0
dtype: int64
Количество дубликатов: 0


**Таблица publishers содержит данные о самом издательстве и его id. В таблице 340 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены.**

In [8]:
query_4 = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
ratings = pd.io.sql.read_sql(sql=text(query_4), con = con)
print('\t\t\t\t Данные о пользовательских оценках книг')
display(ratings)

				 Данные о пользовательских оценках книг


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [9]:
#выведем общую информацию по таблице:
query_4 = '''SELECT * FROM ratings'''
con=engine.connect()
ratings = pd.io.sql.read_sql(sql=text(query_4), con = con)

print ('\n\tИнформация о ratings: \n') 
ratings.info()

#выведем размер исходной таблицы:
print ('\nКоличество строк и столбцов:', ratings.shape)

#проверим на пропуски
print ('\nИнформация по пропускам: \n') 
print(ratings.isnull().sum())

#проверим на дубликаты
print ('Количество дубликатов:', ratings.duplicated().sum())


	Информация о ratings: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB

Количество строк и столбцов: (6456, 4)

Информация по пропускам: 

rating_id    0
book_id      0
username     0
rating       0
dtype: int64
Количество дубликатов: 0


**Таблица ratings содержит данные об имени пльзователя, оставившего оценку, непосредственно саму оценку, id оценки и книги. В таблице 6456 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены.**

In [10]:
query_5 = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
reviews = pd.io.sql.read_sql(sql=text(query_5), con = con)
print('\t\t\t\t Данные о пользовательских обзорах')
display(reviews)

				 Данные о пользовательских обзорах


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [11]:
#выведем общую информацию по таблице:
query_5 = '''SELECT * FROM reviews'''
con=engine.connect()
reviews = pd.io.sql.read_sql(sql=text(query_5), con = con)

print ('\n\tИнформация о reviews: \n') 
reviews.info()

#выведем размер исходной таблицы:
print ('\nКоличество строк и столбцов:', reviews.shape)

#проверим на пропуски
print ('\nИнформация по пропускам: \n') 
print(reviews.isnull().sum())

#проверим на дубликаты
print ('Количество дубликатов:', reviews.duplicated().sum())


	Информация о reviews: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB

Количество строк и столбцов: (2793, 4)

Информация по пропускам: 

review_id    0
book_id      0
username     0
text         0
dtype: int64
Количество дубликатов: 0


**Таблица reviews содержит данные об имени автора обзора, сам текст обзора, id обзора и книги. В таблице 2793 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены.**

**Таким образом, мы исследовали таблицы и можем сделать следующие выводы:**

- **Таблица books содержит данные о названии книги, количестве страниц, дате публикации, id книги, автора и издательства. В таблице 1000 строк и 6 столбцов. Дубликаты и пропущенные значения не обнаружены.**

- **Таблица authors содержит данные об авторе и его id. В таблице 636 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены.**

- **Таблица publishers содержит данные о самом издательстве и его id. В таблице 340 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены.**

- **Таблица ratings содержит данные об имени пльзователя, оставившего оценку, непосредственно саму оценку, id оценки и книги. В таблице 6456 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены.**

- **Таблица reviews содержит данные об имени автора обзора, сам текст обзора, id обзора и книги. В таблице 2793 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены.**

### Посчитать, сколько книг вышло после 1 января 2000 года

In [12]:
a = '''
        SELECT COUNT(DISTINCT book_id ) 
        FROM books 
        WHERE publication_date > '2000-01-01'
    '''
con=engine.connect()
abc = pd.io.sql.read_sql(sql=text(a), con = con)
print('Количество выпущенных книг после 01 января 2000 г.:')
abc

Количество выпущенных книг после 01 января 2000 г.:


count
0    819

**После 01 января 2000 г было выпущено 819 книг.**

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [13]:
b = '''
        SELECT 
               title, 
               bo.book_id,
               COUNT (DISTINCT re.review_id) as count_review_id,
               AVG(ra.rating)  as mean_rating
        FROM books AS bo   
        FULL JOIN reviews AS re ON  bo.book_id = re.book_id
        FULL JOIN ratings AS ra ON  bo.book_id = ra.book_id
        GROUP BY bo.book_id
        ORDER BY count_review_id DESC, mean_rating DESC
    '''
con=engine.connect()
fde = pd.io.sql.read_sql(sql=text(b), con = con)
print('Информация по книгам с количеством обзоров и средним рейтингом')
fde

Информация по книгам с количеством обзоров и средним рейтингом


title  book_id  \
0                              Twilight (Twilight  #1)      948   
1    Harry Potter and the Prisoner of Azkaban (Harr...      302   
2    Harry Potter and the Chamber of Secrets (Harry...      299   
3                                       The Book Thief      656   
4                                     The Glass Castle      734   
..                                                 ...      ...   
995  Disney's Beauty and the Beast (A Little Golden...      191   
996                               Leonardo's Notebooks      387   
997                          Essential Tales and Poems      221   
998    Anne Rice's The Vampire Lestat: A Graphic Novel       83   
999                            The Natural Way to Draw      808   

     count_review_id  mean_rating  
0                  7     3.662500  
1                  6     4.414634  
2                  6     4.287500  
3                  6     4.264151  
4                  6     4.206897  
..               ...          ...  
995                0     4.000000  
996                0     4.000000  
997                0     4.000000  
998                0     3.666667  
999                0     3.000000  

[1000 rows x 4 columns]

**Наибольшее количество обзоров у Twilight (7), а средний рейтинг у Harry Potter and the Prisoner of Azkaban (4.414634).**

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так  исключим из анализа брошюры

In [14]:
c = '''
        SELECT 
              pu.publisher,
              COUNT(DISTINCT bo.book_id) AS count_pub
        FROM books bo
        INNER JOIN publishers AS pu ON pu.publisher_id=bo.publisher_id
        WHERE num_pages > 50
        GROUP BY pu.publisher
        ORDER BY count_pub DESC
        LIMIT 1
    '''
con=engine.connect()
gh = pd.io.sql.read_sql(sql=text(c), con = con)
print('Издательство, выпустившее наибольшее число книг толще 50 страниц')
gh

Издательство, выпустившее наибольшее число книг толще 50 страниц


publisher  count_pub
0  Penguin Books         42

**Наибольшее число книг толще 50 страниц выпустило Penguin Books (42 книги).**

### Определить автора с самой высокой средней оценкой книг — необходимо учитывать только книги с 50 и более оценками

In [15]:
e = '''
        SELECT au.author_id,
               au.author,
               AVG(rating) AS mean_rating
        FROM books bo
        INNER JOIN authors AS au ON au.author_id=bo.author_id
        INNER JOIN ratings AS ra ON ra.book_id=bo.book_id
        WHERE bo.book_id IN (SELECT book_id                  
                          FROM ratings
                          GROUP BY book_id
                          HAVING COUNT(rating_id) >= 50)
        GROUP BY au.author_id                      
        ORDER BY mean_rating DESC
        LIMIT 1
    '''
con=engine.connect()
jrt = pd.io.sql.read_sql(sql=text(e), con = con)
print('Автор с самой высокой средней оценкой книги')
jrt

Автор с самой высокой средней оценкой книги


author_id                      author  mean_rating
0        236  J.K. Rowling/Mary GrandPré     4.287097

**Автор с самой высокой средней оценкой книги - J.K. Rowling/Mary GrandPré (4.287097)**

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [16]:
f = ''' 
        SELECT AVG(count)
        FROM 
            (SELECT username,
                    COUNT (review_id) AS count
             FROM reviews
             WHERE username IN (SELECT username
                               FROM ratings
                               GROUP BY username
                               HAVING COUNT (rating_id) > 48)
            GROUP BY username) AS k           
    '''
con=engine.connect()
jkl = pd.io.sql.read_sql(sql=text(f), con = con)
print('Среднее количество обзоров от пользователей, которые поставили больше 48 оценок')
jkl

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок


avg
0  24.0

**Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.**

### Вывод

**Таким образом, можем сделать следующие выводы:**

- исследовали имеющиеся в нашем распоряжении таблицы:

        - таблица books содержит данные о названии книги, количестве страниц, дате публикации, id книги, автора и издательства. В таблице 1000 строк и 6 столбцов. Дубликаты и пропущенные значения не обнаружены;

        - таблица authors содержит данные об авторе и его id. В таблице 636 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены.

        - таблица publishers содержит данные о самом издательстве и его id. В таблице 340 строк и 2 столбца. Дубликаты и пропущенные значения не обнаружены;

        - таблица ratings содержит данные об имени пльзователя, оставившего оценку, непосредственно саму оценку, id оценки и книги. В таблице 6456 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены;

        - таблица reviews содержит данные об имени автора обзора, сам текст обзора, id обзора и книги. В таблице 2793 строки и 4 столбца. Дубликаты и пропущенные значения не обнаружены.
        
- После 01 января 2000 г было выпущено 819 книг.

- Наибольшее количество обзоров у Twilight (7), а средний рейтинг у Harry Potter and the Prisoner of Azkaban (4.414634).

- Наибольшее число книг толще 50 страниц выпустило Penguin Books (42 книги).

- Автор с самой высокой средней оценкой книги - J.K. Rowling/Mary GrandPré (4.287097)

- Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.